Bluetooth: host: Fix issues in use of big-handle from HCI events

joerchan opened the pull request:
Fix parsing BIG handle which is just one octet as a two octet field,
which could have caused problems on big-endian systems.

Validate the big_handle before received from the controller before
using it as an array into the bigs array.

Discover by coverity: Insecure data handling  (TAINTED_SCALAR)
  Using tainted variable "big_handle" as an index into an array "bigs".

Fixes: #33312
Fixes: #33313

joerchan committed:
Bluetooth: host: Fix parsing BIG handle as a two byte field

Fix parsing BIG handle which is just one octet as a two octet field,
which could have caused problems on big-endian systems.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Validate big_handle before using it as an array index

Validate the big_handle before received from the controller before
using it as an array into the bigs array.

Discover by coverity: Insecure data handling  (TAINTED_SCALAR)
  Using tainted variable "big_handle" as an index into an array "bigs".

Cleanup state of BIG in pending or syncing state if it exists.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1383,7 +1383,14 @@ int bt_iso_big_terminate(struct bt_iso_big *big)
-	struct bt_iso_big *big = &bigs[evt->big_handle];
+	struct bt_iso_big *big;
+
+	if (evt->big_handle > ARRAY_SIZE(bigs)) {
```
Shouldn't this be `>=` (and in all the other places as well)?

joerchan replied:
Yes, fixed that.

Thalley replied:
@joerchan while your new change works, isn't it simpler to just do as suggested by Johan and do `>=` rather than `! <`?

joerchan replied:
Sure, I just fixed this before I saw Johans comment.

I guess this is subjective, but in my head i find it easier to deal with negations of the expression rather than the opposite expression (as you can see by the fact I messed this up :P).
I can change to `>=` if that is preferred.

jhedberg replied:
`>=` is easier to read for me at least

jhedberg replied:
@jhedberg wrote:
> \>= is easier to read for me at least

@joerchan meaning, I'll wait for you to update to that before approving 😉

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1417,8 +1430,14 @@ void hci_le_big_terminate(struct net_buf *buf)
-	struct bt_iso_big *big = &bigs[big_handle];
+	struct bt_iso_big *big;
+
+	if (evt->big_handle > ARRAY_SIZE(bigs)) {
+		BT_WARN("Invalid BIG handle");
+		return;
+	}
```
Similar here; we should terminate the pending BIG

joerchan commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1383,7 +1383,14 @@ int bt_iso_big_terminate(struct bt_iso_big *big)
+	if (evt->big_handle > ARRAY_SIZE(bigs)) {
+		BT_WARN("Invalid BIG handle");
+		return;
+	}
```
I suppose you are right, you know this better than me.

How do I do that?

Thalley replied:
I guess we first need to find a way to mark a BIG as pending (there can only be a single BIG initiating or a single BIG syncing at one time (although we can have a BIG initiating and another BIG syncing at the same time), so I suggest adding a (atomic?) flag to the BIG struct with a `syncing` and a `initiating` bit and when this even arrives with an invalid handle, we iterate the array of BIGs and find the syncing/initiating (based on the event) and terminate it. 

To terminate an invalid BIG we would probably need to change the `struct bt_iso_big *big` parameter of `hci_le_terminate_big` and `hci_le_big_sync_term` to just be the (invalid) handle, and then also call `big_disconnect(big);` and `cleanup_big(big);` with the BIG we looked up so that we clean up the host as well as notifying the application.

joerchan replied:
Is it correct now, I haven't done `big_disconnect` because other error handling in this function also did not do this. Is this something that needs to be improved?

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/33461#discussion_r597510476

joerchan committed:
Bluetooth: host: Refactor BIG initialized handling to use bit flags

Refactor the BIG initialized atomic handling to use bits instead of a
single atomic.
This allows for more bit flags to be added.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1140,7 +1153,9 @@ static void cleanup_big(struct bt_iso_big *big)
 	big->bis = NULL;
 	big->num_bis = 0;
-	atomic_clear(&big->initialized);
+	atomic_clear_bit(big->flags, BT_BIG_SYNCING);
+	atomic_clear_bit(big->flags, BT_BIG_PENDING);
+	atomic_clear_bit(big->flags, BT_BIG_INITIALIZED);
```
```suggestion
	atomic_clear(&big->flags);
```

joerchan replied:
It's not the same type:
```
warning: passing argument 1 of 'atomic_clear' from incompatible pointer type [-Wincompatible-pointer-types]
 1156 |  atomic_clear(&big->flags);
 ```
 
 Could update to a memset of the struct?

Thalley replied:
Might just be `atomic_clear(big->flags);`. We could memset the entire struct as well as that's what we are basically doing anyhow.

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1383,7 +1398,21 @@ int bt_iso_big_terminate(struct bt_iso_big *big)
+
+		big = big_lookup_flag(BT_BIG_PENDING);
+		if (big) {
+			cleanup_big(big);
```
I think we need to do the same here as we do for per-adv and call the disconnected callbacks by calling `big_disconnect` before `cleanup_big`.

joerchan replied:
I'm unsure about the `bit_disconnect` you mentioned, if you see below here other if the number of BIS is invalid there is also just cleanup_big.

Is this correct?

Thalley replied:
I think that's actually rather a missing `big_disconnect(big);` from below. I guess I was unsure whether or not to call the connected/disconnect callbacks for the broadcaster, but since we are also calling `bt_conn_set_state(bis->conn, BT_CONN_CONNECTED);` (which calls the `connected` callback), then we should also call the `disconnected` callback. 

So I suggest adding it to both places.

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1127,6 +1127,19 @@ static struct bt_iso_big *get_free_big(void)
+static struct bt_iso_big *big_lookup_flag(int bit)
+{
+	for (int i = 0; i < ARRAY_SIZE(bigs); i++) {
+		if (!atomic_test_bit(bigs[i].flags, bit)) {
```
```suggestion
		if (atomic_test_bit(bigs[i].flags, bit)) {
```

joerchan replied:
Thank you for testing, fixed.

Thalley reviewed (changes requested):
A few responses

joerchan committed:
Bluetooth: host: Fix missing big_disconnect when invalid number of BIS

Fix missing big_disconnect in hci_le_big_complete when the number of
BIS in the event is invalid.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
